### PR TITLE
Replaced WebUtility.HtmlEncode with IHtmlEncoder.HtmlEncode

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #304 
@dougbu @NTaylorMullen

### CODE REVIEW

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 1]
```diff
@@ -33,12 +34,18 @@ internal TagHelperOutput(string tagName)
         /// <param name="tagName">The HTML element's tag name.</param>
         /// <param name="attributes">The HTML attributes.</param>
         public TagHelperOutput(string tagName, [NotNull] IDictionary<string, string> attributes)
+            : this(tagName, attributes, HtmlEncoder.Default)
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: @dougbu @NTaylorMullen, Please decide if this is needed.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 2]
```diff
@@ -33,12 +34,18 @@ internal TagHelperOutput(string tagName)
...
+        {
+        }
+
+        public TagHelperOutput(string tagName, [NotNull] IDictionary<string, string> attributes, IHtmlEncoder htmlEncoder)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: wrap long line

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: I prefer no.  if tag helper authors create `TagHelperOutput`s, they'll have easy access (`[Activate]`) to an `IHtmlEncoder` implementation.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs [hunk 3]
```diff
@@ -26,7 +34,7 @@ public class TagHelperRunner
                 executionContext.Items,
                 executionContext.UniqueId,
                 executionContext.GetChildContentAsync);
-            var tagHelperOutput = new TagHelperOutput(executionContext.TagName, executionContext.HTMLAttributes)
+            var tagHelperOutput = new TagHelperOutput(executionContext.TagName, executionContext.HTMLAttributes, _htmlEncoder)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs]: wrap long line

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs [hunk 4]
```diff
@@ -48,7 +48,13 @@ protected override void Visit(TagHelperChunk chunk)
...
                     WritePrivateField(_tagHelperContext.RunnerTypeName,
                                       CSharpTagHelperCodeRenderer.RunnerVariableName,
-                                      "new " + _tagHelperContext.RunnerTypeName + "()");
+                                      "new "
+                                      + _tagHelperContext.RunnerTypeName
+                                      + "("
+                                      + _tagHelperContext.HtmlHelperPropertyName
+                                      + "."
+                                      + _tagHelperContext.HtmlEncoderPropertyName
+                                      + ")");
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs]: too many concatenations.  use the `Writer` directly and pass each string individually.

File: src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs [hunk 5]
```diff
@@ -35,6 +35,8 @@ public GeneratedTagHelperContext()
             RunnerTypeName = "TagHelperRunner";
             ScopeManagerTypeName = "TagHelperScopeManager";
             ExecutionContextTypeName = "TagHelperExecutionContext";
+            HtmlEncoderPropertyName = "HtmlEncoder";
+            HtmlHelperPropertyName = "Html";
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs]: nit: don't really need two properties and this limits what `MvcRazorHost` can override.  just `HtmlEncoderPropertyName` with a default of `Html.HtmlEncoder` should be fine unless @NTaylorMullen objects.

File: src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs [hunk 6]
```diff
@@ -150,5 +152,15 @@ public GeneratedTagHelperContext()
         /// Contains information about in-scope tag helpers, HTML attributes, and the tag helpers' output.
         /// </remarks>
         public string ExecutionContextTypeName { get; set; }
+
+        /// <summary>
+        /// The name of the property in HtmlHelper used to encode html.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs]: - "name of the property" -> "property name"
- "html" -> "HTML"

Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs]: better if `HtmlHelperPropertyName` remains: `The name of the <c>IHtmlHelper</c> property ...`.  Otherwise `The name of the property used to encode HTML.  Expected to be of type <c>IHtmlEncoder</c>.`

File: src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs [hunk 7]
```diff
@@ -150,5 +152,15 @@ public GeneratedTagHelperContext()
...
+        /// </summary>
+        public string HtmlEncoderPropertyName { get; set; }
+
+        /// <summary>
+        /// The name of the injected property of type IHtmlHelper.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs]: if this remains don't mention "injected" -- that's up to MVC to decide.  suggest `The name of the property containing <see cref="HtmlEncoderPropertyName"/>.`  (hmm, mentioning `IHtmlHelper` with `HtmlEncoderPropertyName` isn't worthwhile)

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperRunnerTest.cs [hunk 8]
```diff
@@ -61,7 +62,7 @@ public static TheoryData TagHelperOrderData
             int[] expectedTagHelperOrders)
         {
             // Arrange
-            var runner = new TagHelperRunner();
+            var runner = new TagHelperRunner(new HtmlEncoder());
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperRunnerTest.cs]: what's the issue with using a `TestHtmlEncoder` class right now?  basically just filed aspnet/Mvc#2085 and now extending that problem into the Razor code base instead of writing a 20 line private class at the bottom of this one.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: I'd rather make the API simple (I like it as is), only time users will care about providing an HTMLEncoder is if they've already registered one themselves therefore it makes sense for them to pull in their own and provide it.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 9]
```diff
@@ -19,6 +19,7 @@ public class TagHelperOutput
         private bool _contentSet;
         private bool _isTagNameNullOrWhitespace;
         private string _tagName;
+        private IHtmlEncoder _htmlEncoder;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: readonly

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: docs

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs [hunk 10]
```diff
@@ -12,6 +13,13 @@ namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
     /// </summary>
     public class TagHelperRunner
     {
+        private IHtmlEncoder _htmlEncoder;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs]: readonly

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs [hunk 11]
```diff
@@ -12,6 +13,13 @@ namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
...
+
+        public TagHelperRunner(IHtmlEncoder htmlEncoder)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs]: [NotNull]

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs]: docs

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs]: :+1: , the `WritePrivateField` has 2 writes to it so its not all taht much duplication

File: src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs]: Completely agree, 1 configuration value is best here, Razor doesn't need to know about the Html helper

File: src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs (hunk 7 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs]: Nuke this :smile:

**Comment by NTaylorMullen** [comment]
Add a test validating that `TagHelperOutput` uses the provided encoder & add a test that verifies that the `TagHelperRunner` create `TagHelperOutput`s with the provided encoder.

**Comment by NTaylorMullen** [comment]
:watch:

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: @NTaylorMullen who is "they"?  tag helper authors aren't the people writing `Startup` classes for actual apps.  if we leave this, the authors will break their users.  and we could have prevented it.

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: With how we've designed `TagHelper`s users/authors can be the same; we've made them simple enough to be used as enhancements to the "average" application. For `TagHelper` framework authors such as ourselves if we ever plan to `new` up `TagHelperOutput`s then we'd forcefully set the HTMLEncoder, much like we do for `string.Format` when providing the culture. Ultimately we have the option for when it makes sense otherwise it's still easy.

Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: @NTaylorMullen the `IHtmlHelper` comes from DI, not this `static` property in any real scenario.

my point is users and authors are most often not the same.  us as framework authors creating the MVC tag helpers and our customers is just one example of the separation.  basically what you're making "easy" is creation of a bunch of tag helpers that use an `IHtmlEncoder` implementation that application authors can't override.

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Sure, but in the majority case users will never provide an `IHtmlEncoder` to begin with so why enforce providing it in `TagHelperOutput`s construction? If a `TagHelper` author is serious about the `TagHelper`s they create they'll be explicit as to which encoder they want used.

**Comment by ajaybhargavb** [comment]
Updated.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 12]
```diff
@@ -32,13 +33,17 @@ internal TagHelperOutput(string tagName)
         /// </summary>
         /// <param name="tagName">The HTML element's tag name.</param>
         /// <param name="attributes">The HTML attributes.</param>
-        public TagHelperOutput(string tagName, [NotNull] IDictionary<string, string> attributes)
+        /// <param name="htmlEncoder">The <see cref="IHtmlEncoder"/> to be used to encode HTML.</param>
+        public TagHelperOutput(string tagName,
+                               [NotNull] IDictionary<string, string> attributes,
+                               IHtmlEncoder htmlEncoder)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: `[NotNull]`

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 13]
```diff
@@ -32,13 +33,17 @@ internal TagHelperOutput(string tagName)
         /// </summary>
         /// <param name="tagName">The HTML element's tag name.</param>
         /// <param name="attributes">The HTML attributes.</param>
-        public TagHelperOutput(string tagName, [NotNull] IDictionary<string, string> attributes)
+        /// <param name="htmlEncoder">The <see cref="IHtmlEncoder"/> to be used to encode HTML.</param>
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Isn't used for everything, only used to encode attribute values.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs [hunk 14]
```diff
@@ -26,7 +38,9 @@ public class TagHelperRunner
                 executionContext.Items,
                 executionContext.UniqueId,
                 executionContext.GetChildContentAsync);
-            var tagHelperOutput = new TagHelperOutput(executionContext.TagName, executionContext.HTMLAttributes)
+            var tagHelperOutput = new TagHelperOutput(executionContext.TagName,
+                                                      executionContext.HTMLAttributes,
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs]: nit: don't add the extra spacing, just let it lie 1 indent in (that's where it goes when you just hit enter on the previous line.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 15]
```diff
@@ -32,13 +33,17 @@ internal TagHelperOutput(string tagName)
...
+        public TagHelperOutput(string tagName,
+                               [NotNull] IDictionary<string, string> attributes,
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: https://github.com/aspnet/Razor/pull/314/files#r25538143

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs [hunk 16]
```diff
@@ -46,13 +46,20 @@ protected override void Visit(TagHelperChunk chunk)
...
-                                      "new " + _tagHelperContext.RunnerTypeName + "()");
+                    Writer.Write("private ")
+                          .Write(_tagHelperContext.RunnerTypeName)
+                          .Write(" ")
+                          .WriteStartAssignment(CSharpTagHelperCodeRenderer.RunnerVariableName)
+                          .WriteStartNewObject(_tagHelperContext.RunnerTypeName)
+                          .Write(_tagHelperContext.HtmlEncoderPropertyName)
+                          .WriteEndMethodInvocation();
 
-                    WritePrivateField(_tagHelperContext.ScopeManagerTypeName,
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs]: This one is fine as is, undo these changes.

File: src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs [hunk 17]
```diff
@@ -150,5 +151,10 @@ public GeneratedTagHelperContext()
         /// Contains information about in-scope tag helpers, HTML attributes, and the tag helpers' output.
         /// </remarks>
         public string ExecutionContextTypeName { get; set; }
+
+        /// <summary>
+        /// The name of the property used to encode HTML. Expected to be of type <c>IHtmlEncoder</c>.
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/GeneratedTagHelperContext.cs]: Remove `Expected to be of type <c>IHtmlEncoder</c>.`, depending on how they configure the `GeneratedTagHelperContext` it may not be a requirement.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs [hunk 18]
```diff
@@ -79,13 +79,14 @@ public void GenerateStartTag_ReturnsFullStartTag()
                 {
                     { "class", "btn" },
                     { "something", "   spaced    " }
-                });
+                },
+                htmlEncoder: new TestHtmlEncoder());
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: I don't think we should be touching existing tests and providing odd encoders. The default encoder is fine.

**Comment by NTaylorMullen** [comment]
Close, :watch:

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs (hunk 18 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: We don't want to use the default encoder for unit testing. https://github.com/aspnet/Mvc/pull/2076#discussion-diff-25388824

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs (hunk 16 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs]: @dougbu wanted me to change this as well.

**Comment by ajaybhargavb** [comment]
Updated.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 19]
```diff
@@ -32,13 +33,18 @@ internal TagHelperOutput(string tagName)
         /// </summary>
         /// <param name="tagName">The HTML element's tag name.</param>
         /// <param name="attributes">The HTML attributes.</param>
-        public TagHelperOutput(string tagName, [NotNull] IDictionary<string, string> attributes)
+        /// <param name="htmlEncoder">The <see cref="IHtmlEncoder"/> to be used
+        /// to encode HTML attribute values.</param>
+        public TagHelperOutput(string tagName,
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Move this with the other ctor params below, basically just hit enter.

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs [hunk 20]
```diff
@@ -26,7 +38,9 @@ public class TagHelperRunner
                 executionContext.Items,
                 executionContext.UniqueId,
                 executionContext.GetChildContentAsync);
-            var tagHelperOutput = new TagHelperOutput(executionContext.TagName, executionContext.HTMLAttributes)
+            var tagHelperOutput = new TagHelperOutput(executionContext.TagName,
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs]: https://github.com/aspnet/Razor/pull/314/files#r25544145

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs (hunk 16 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs]: argh, alright.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs (hunk 18 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: Talked in person

**Comment by NTaylorMullen** [comment]
:watch: just want to see the testing changes.

**Comment by ajaybhargavb** [comment]
Updated.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TestHtmlEncoder.cs [hunk 21]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class TestHtmlEncoder : IHtmlEncoder
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TestHtmlEncoder.cs]: suggest naming this a bit more clearly e.g. `PassThroughHtmlEncoder`

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/CustomHtmlEncoder.cs [hunk 22]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class CustomHtmlEncoder : IHtmlEncoder
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/CustomHtmlEncoder.cs]: suggest `PseudoHtmlEncoder`

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TestHtmlEncoder.cs (hunk 21 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TestHtmlEncoder.cs]: or `NullHtmlEncoder` or `NoOpHtmlEncoder`

**Comment by dougbu** [comment]
suggest pulling at least the pass-through HTML encoder into the Testing repo as the first step of the aspnet/MVC#2085 fix.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs [hunk 23]
```diff
@@ -137,6 +139,26 @@ public void GenerateStartTag_ReturnsSelfClosingStartTag_NoAttributes()
...
         [Fact]
+        public void GenerateStartTag_UsesProvidedHtmlEncoder()
+        {
+            // Arrange
+            var tagHelperOutput = new TagHelperOutput("p",
+                attributes: new Dictionary<string, string>
+                {
+                    { "hello", "world" },
+                },
+                htmlEncoder: new CustomHtmlEncoder());
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: Love it, beauuutiful! :+1:

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TestHtmlEncoder.cs (hunk 21 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TestHtmlEncoder.cs]: `NullHtmlEncoder` falls same format as the logging bits that just went into testing.

**Comment by NTaylorMullen** [comment]
:shipit: once @dougbu is happy.

**Comment by ajaybhargavb** [comment]
Passing `IHtmlEncoder` to `TagHelperRunner` doesn't work. It throws this error (A field initializer cannot reference the non-static field, method, or property 'RazorPage.HtmlEncoder') when trying to create a `private TagHelperRunner __tagHelperRunner = new TagHelperRunner(HtmlEncoder)`.

**Comment by NTaylorMullen** [comment]
I suggest we have Razor create an activate property for the taghelper runner.

**Comment by dougbu** [comment]
@NTaylorMullen good idea.  another way of doing it be something like the following (every time).  activation seems better.  (requires a change from using `CSharpTagHelperCodeRenderer.RunnerVariableName`, `GeneratedTagHelperContext.RunnerTypeName`, and not writing out the declaration to something like `GeneratedTagHelperContext.RunnerPropertyName`)

``` c#
__tagHelperRunner = __tagHelperRunner ?? new TagHelperRunner(HtmlEncoder);
__tagHelperExecutionContext.Output = __tagHelperRunner.RunAsync(__tagHelperExecutionContext).Result;
```

**Comment by NTaylorMullen** [comment]
Actuallly shoot, I've talked myself out of the `[Activate]` approach because that'd require us to include the `TagHelperRunner` in DI which is somewhat odd since none of the other `TagHelper` bits are in DI. How about a mix of the two and have the `__tagHelperRunner` be instantiated at the top of the `ProcessAsync` method?

```C#
public Task ProcessAsync()
{
    __tagHelperRunner = new TagHelperRunner(HtmlEncoder);
    ...
}
```

**Comment by dougbu** [comment]
@NTaylorMullen `[Activate]` works for services, yes.  but it also works for anything that we wish to special-case in `DefaultTagHelperActivator`.  so this case could work using `[Activate]`.  the alternate you suggest works as well but involves `RazorPage` methods directly referencing generated fields.  let's discuss tomorrow.

**Comment by ajaybhargavb** [comment]
Updated. @NTaylorMullen @dougbu

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/NullHtmlEncoder.cs [hunk 24]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/NullHtmlEncoder.cs]: `.Test`, it's a helper.

File: test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/PseudoHtmlEncoder.cs [hunk 25]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/PseudoHtmlEncoder.cs]: ditto

**Comment by NTaylorMullen** [comment]
:shipit:

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs [hunk 26]
```diff
@@ -46,13 +46,18 @@ protected override void Visit(TagHelperChunk chunk)
...
                                       value: null);
 
-                    WritePrivateField(_tagHelperContext.RunnerTypeName,
-                                      CSharpTagHelperCodeRenderer.RunnerVariableName,
-                                      "new " + _tagHelperContext.RunnerTypeName + "()");
+                    Writer.Write("private ")
+                          .WriteVariableDeclaration(
+                        _tagHelperContext.RunnerTypeName,
+                        CSharpTagHelperCodeRenderer.RunnerVariableName,
+                        value: null);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs]: nit: very strange indentation.  hit <Enter> after `Writer` and let VS line the rest up.

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperRunnerInitializationVisitor.cs [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+            if (!_foundTagHelpers && !Context.Host.DesignTimeMode)
+            {
+                _foundTagHelpers = true;
+
+                Writer.WriteStartAssignment(CSharpTagHelperCodeRenderer.RunnerVariableName)
+                      .Write(CSharpTagHelperCodeRenderer.RunnerVariableName)
+                      .Write(" ?? ")
+                      .WriteStartNewObject(_tagHelperContext.RunnerTypeName)
+                      .Write(_tagHelperContext.HtmlEncoderPropertyName)
+                      .WriteEndMethodInvocation();
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperRunnerInitializationVisitor.cs]: FYI (no need to change anything in this PR) it's much easier and helps future developers if you just let VS do its thing.  in this case that would look more like
``` c#
Writer
    .WriteStartAssignment(CSharpTagHelperCodeRenderer.RunnerVariableName)
    .Write(CSharpTagHelperCodeRenderer.RunnerVariableName)
    ...;
```

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 28]
```diff
@@ -32,13 +33,19 @@ internal TagHelperOutput(string tagName)
         /// </summary>
         /// <param name="tagName">The HTML element's tag name.</param>
         /// <param name="attributes">The HTML attributes.</param>
-        public TagHelperOutput(string tagName, [NotNull] IDictionary<string, string> attributes)
+        /// <param name="htmlEncoder">The <see cref="IHtmlEncoder"/> to be used
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: teensy nit: don't need "to be" (not one of my better suggestions if that was mine)

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperRunnerInitializationVisitor.cs [hunk 29]
```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNet.Razor.Generator.Compiler.CSharp
+{
+    public class CSharpTagHelperRunnerInitializationVisitor : CodeVisitor<CSharpCodeWriter>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperRunnerInitializationVisitor.cs]: doc comments for class and its members

**Comment by dougbu** [comment]
:watch: just for the `CSharpTagHelperRunnerInitializationVisitor` documentation

**Comment by ajaybhargavb** [comment]
Updated. @dougbu

File: src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs [hunk 30]
```diff
@@ -12,6 +13,17 @@ namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
     /// </summary>
     public class TagHelperRunner
     {
+        private readonly IHtmlEncoder _htmlEncoder;
+
+        /// <summary>
+        /// Instantiates a new instance of <see cref="TagHelperRunner"/>.
+        /// </summary>
+        /// <param name="htmlEncoder">The <see cref="IHtmlEncoder"/> to be used to encode HTML.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperRunner.cs]: nit: previous "to be" nit was general

File: src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperRunnerInitializationVisitor.cs [hunk 31]
```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNet.Razor.Generator.Compiler.CSharp
+{
+    /// <summary>
+    /// The CodeVisitor that generates the code to initialize the TagHelperRunner.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/Visitors/CSharpTagHelperRunnerInitializationVisitor.cs]: CodeVisitor -> `<see cref="CodeVisitor{T}"/>`

**Comment by dougbu** [comment]
:shipit: with one small change

**Comment by NTaylorMullen** [comment]
Nuke branch when you're ready.
